7.2. - Tabulating and cross-tabulating data

- A construçãao de tabelas de frequência ou tabulações cruzadas de uma variável por outra variável é bastante comum em análises de dados

- Para construir uma tabela de frequências basta fazer `pd.crosstab(index = df["coluna"], columns = "count")` -> Gera  uma tabela de frequência com a quantidade total de cada elemento da coluna

- O código abaixo gera um DataFraame genérico de carros e suas respectivas cores

In [1]:
import pandas as pd

data = {
    "carro": ["Ford", "Fiat", "Ford", "Renault", "BMW", "Fiat", "Fiat", "BMW"],
    "cor": ["preto", "preto", "vermelho", "branco", "azul", "preto", "preto", "prata"]
}

df = pd.DataFrame(data, columns=["carro", "cor"])

df

,carro,cor
0,Ford,preto
1,Fiat,preto
2,Ford,vermelho
3,Renault,branco
4,BMW,azul
5,Fiat,preto
6,Fiat,preto
7,BMW,prata


- Para realizar a contagem da quantidade total de carros por marca basta utilizar `pd.crosstab()`

In [3]:
pd.crosstab(index = df["carro"], columns = "total")

col_0,total
carro,
BMW,2
Fiat,3
Ford,2
Renault,1


- No código anterior a coluna de contagem foi definida genericamente como 'total', mas também seria possível gerar uma tabela de frequência de carros pela cor

In [4]:
pd.crosstab(index = df["carro"], columns = df["cor"], margins=True)

cor,azul,branco,prata,preto,vermelho,All
carro,,,,,,
BMW,1,0,1,0,0,2
Fiat,0,0,0,3,0,3
Ford,0,0,0,1,1,2
Renault,0,1,0,0,0,1
All,1,1,1,4,1,8


7.2.1. - Converting a table of counts to a table of proportions

- Os comandos anteriores constroem tabelas de frequências simples que são úteis apenas em alguns casos de situações específicas, não atendendo aos casso em que é necessário uma tabela de proporção e não de contagem total

- Para diferenciar a quantidade total de linhas e colunas é possível renomar a linha e coluna chamada de 'All' do código anterior para refletir melhor o dado armazenado

In [10]:
tabelas = pd.crosstab(index=df["carro"], columns=df["cor"], margins=True)

tabelas.columns = list(tabelas.columns)[0:-1] + ["linhas_totais"]
tabelas.index = list(tabelas.index)[0:-1] + ["colunas_totais"]

tabelas

,azul,branco,prata,preto,vermelho,linhas_totais
BMW,1,0,1,0,0,2
Fiat,0,0,0,3,0,3
Ford,0,0,0,1,1,2
Renault,0,1,0,0,0,1
colunas_totais,1,1,1,4,1,8


- `list(df)` ou `df.columns` -> Retornam uma lista de todas as tabela do DataFrame

- `list(df.columns)` -> Converte as colunas do DatFrme para uma lista, permitindo todas as manipulações padrões com listas como substituir itens. O mesmo se aplica para `df.index`

- Para gerar uma tabela de proporção basta dividir o DataFrame de frequência pelo total em cada coluna

In [11]:
tabelas/tabelas.loc["colunas_totais"]

,azul,branco,prata,preto,vermelho,linhas_totais
BMW,1.0,0.0,1.0,0.00,0.0,0.250
Fiat,0.0,0.0,0.0,0.75,0.0,0.375
Ford,0.0,0.0,0.0,0.25,1.0,0.250
Renault,0.0,1.0,0.0,0.00,0.0,0.125
colunas_totais,1.0,1.0,1.0,1.00,1.0,1.000


- A tabela anterior revela que:
  - Por marca de carro:
    1. Fiat tem a maior participação com 37.5% dos carros
    2. BMW e Fort empatam com 25% dos carros
    3. Renault tem a menor participação com 12.5%

  - Por cor de carro:
    1. Todas as cores têm exatamente 1 carro cada (1.0), ou seja, a distribuição de cores é uniforme

- O procedimento para obter a proporção por linha é um pouco diferente:

In [13]:
tabelas.div(tabelas["linhas_totais"], axis=0)

,azul,branco,prata,preto,vermelho,linhas_totais
BMW,0.500,0.000,0.500,0.0,0.000,1.0
Fiat,0.000,0.000,0.000,1.0,0.000,1.0
Ford,0.000,0.000,0.000,0.5,0.500,1.0
Renault,0.000,1.000,0.000,0.0,0.000,1.0
colunas_totais,0.125,0.125,0.125,0.5,0.125,1.0


7.3. - Transforming and recording a variable

- No cenário real de análise de dados é comum que a variável trabalhada não seja extamente aquilo que é necessário, nesses casos é comum a conversão de dados de valores contínuos (como idade) em outras variáveis menores (como criança, jovem, adulto, etc).

7.3.1. - Creating a transformed variable

- Transformar uma variável significa que será aplicada uma função matemática simples na variável original a fim de criar um novo dado que descreva melhor o que é necessário e que esteja mais próxima do que se deseja para realizar testes estatísticos


- O exemplo seguinte foi extraído do tópico [7.3.1 do livro Learning Statistics with Python](https://ethanweed.github.io/pythonbook/03.03-pragmatic_matters.html#creating-a-transformed-variable)

> On a scale of 1 (strongly disagree) to 7 (strongly agree), to what extent do you agree with the proposition that “Dinosaurs are awesome”?

In [27]:
dados = [1, 3, 5, 2, 1, 6, 7, 4, 5, 7]

- O problema é que existe um senso comum em que o ponto central de uma escala deve ser igual a zero (opinião neutra) e os extremos da esquerda e direita representariam a total discordância ou concordância

- Para modificar os dados da pesquisa, basta criar uma equação simples que subtrai 4 de todos os dados coletados

In [28]:
dados = [x-4 for x in dados]
print(dados)

# O mesmo se aplicaria em um array do NumPy
import numpy as np

arr_dados = np.array([1, 3, 5, 2, 1, 6, 7, 4, 5, 7])
arr_dados = arr_dados - 4
print(arr_dados)

[-3, -1, 1, -2, -3, 2, 3, 0, 1, 3]
[-3 -1  1 -2 -3  2  3  0  1  3]


- Realizar essa operação é interessante para os casos em que o impacto (ou força) de cada dado ou a direção a que eles tendem

- Para calcular a força de opinião é necessário calcular o valor absuluto do dado central usando a função abs()

In [33]:
arr_dados = np.array([1, 7, 3, 4, 4, 4, 2, 6, 5, 5])
arr_dados = arr_dados -4
arr_dados = abs(arr_dados)
arr_dados

array([3, 3, 1, 0, 0, 0, 2, 2, 1, 1])

- Para calcular uma variável que contenha apenas a direção da opinião, basta utilizar o método `numpy.sign()`

- O método sign() do numpy converte todos os números negativos em -1 e todos os números positivos são convertidos para 1 (zeros permancem inalterados).

In [34]:
arr_dados = np.array([1, 7, 3, 4, 4, 4, 2, 6, 5, 5])
arr_dados = arr_dados - 4
arr_dados = np.sign(arr_dados)
arr_dados

array([-1,  1, -1,  0,  0,  0, -1,  1,  1,  1])

- Com essas três variáveis transformadas é possível montar um DataFrame para analisar as diferenças

In [35]:
import pandas as pd
df = pd.DataFrame(
    {"pontos": np.array([1, 7, 3, 4, 4, 4, 2, 6, 5, 5])
    })

df["centro"] = df["pontos"]-4
df["forca_opiniao"] = abs(df["centro"])
df["direcao_opiniao"] = np.sign(df["pontos"]-4)
df

,pontos,centro,forca_opiniao,direcao_opiniao
0,1,-3,3,-1
1,7,3,3,1
2,3,-1,1,-1
3,4,0,0,0
4,4,0,0,0
5,4,0,0,0
6,2,-2,2,-1
7,6,2,2,1
8,5,1,1,1
9,5,1,1,1


7.3.2. Cutting a numeric variable into categories

- Em alguns casos é necessário cortar uma variável numérica em categorias discretas, por exemplo:

In [36]:
import pandas as pd
df = pd.DataFrame(
    {"idade": np.array([12, 18, 60, 25, 32, 44, 13, 9, 57, 33])
    })

df

,idade
0,12
1,18
2,60
3,25
4,32
5,44
6,13
7,9
8,57
9,33


- Em alguns casos seria mais útil agrupar essas idades em grupos categóricos menores

- jovens (0-20); adultos (21-40); idosos (41-60)

- Pandas possui o método `.cut()` para o agrupamento de dados

In [37]:
df["categorias"] = pd.cut(x = df["idade"], bins = [0,20,40,60], labels = ["jovem", "adulto", "idoso"])

df

,idade,categorias
0,12,jovem
1,18,jovem
2,60,idoso
3,25,adulto
4,32,adulto
5,44,idoso
6,13,jovem
7,9,jovem
8,57,idoso
9,33,adulto


- No argumento `bins` foram passados 4 valores (0, 20, 40 e 60), mas há somente 3 categorias. Isso é necessário porque o Pandas exige que sejam passados as bordas das categorias ao invés do ponto central

- O método cut possui muitos argumentos, são eles:
  - x: A variável que precisa ser categorizada
  - bins: Um vetor que contêm as localizações dos pontos de separação das categorias ou o número que indica a quantidade de categorias desejadas
  - labels: Os rótulos anexados para cada categoria (opcional)

- O usuário pode deixar que o Python decida por si só como categorizar os dados. O próximo código só irá definir a quantidade de categorias desejadas

In [38]:
df["categorias"] = pd.cut(x = df["idade"], bins = 3)

df

,idade,categorias
0,12,"(8.949, 26.0]"
1,18,"(8.949, 26.0]"
2,60,"(43.0, 60.0]"
3,25,"(8.949, 26.0]"
4,32,"(26.0, 43.0]"
5,44,"(43.0, 60.0]"
6,13,"(8.949, 26.0]"
7,9,"(8.949, 26.0]"
8,57,"(43.0, 60.0]"
9,33,"(26.0, 43.0]"


- Todas as informação importantes podem ser extraídas olhando os dados tabulados

- O código abaixo irá exibir a quantidade total de itens por categoria

In [40]:
pd.crosstab(index = df["categorias"], columns = "total")

col_0,total
categorias,
"(8.949, 26.0]",5
"(26.0, 43.0]",2
"(43.0, 60.0]",3


- O que o Python faz é tentar dividir as idades em 3 categorias de bins com tamanhos iguais

- Em alguns casos será necessário em categorias de tamanhos diferentes, mas com um número idêntico de casos, para essas situações será necessário adicionar os limites correspondentes em porcentagem

- Utilizando o método `qcut()` que divide por quantidades iguais em percentis, gerando grupos de tamanhos diferentes, mas com quantidades iguais

- Ou seja:
  - `cut()` -> Corta os dados em pedaços de tamanhos fixos
  - `qcut()` -> Corta os dados para que cada grupo tenha o mesmo número de elementos (divide em grupos com quantidade de itens equilibrados)

In [42]:
df["categorias"] = pd.qcut(x = df["idade"], q = [0, .33, .66, 1])

df

,idade,categorias
0,12,"(8.999, 17.85]"
1,18,"(17.85, 32.94]"
2,60,"(32.94, 60.0]"
3,25,"(17.85, 32.94]"
4,32,"(17.85, 32.94]"
5,44,"(32.94, 60.0]"
6,13,"(8.999, 17.85]"
7,9,"(8.999, 17.85]"
8,57,"(32.94, 60.0]"
9,33,"(32.94, 60.0]"
